src/github.com/getlantern/flashlight/ui/service.go

```diff
@@ -89,14 +114,29 @@ func read() {
 		}
 
 		// Delegating response to the service that registered with the given type.
-		if services[env.Type] != nil {
-			// Pass this message and continue reading another one.
-			go func() {
-				services[env.Type].In <- b
-			}()
-		} else {
-			log.Errorf("Message type %s belongs to an unkown service.", env.Type)
+		if services[envType.Type] == nil {
+			log.Errorf("Message type %v belongs to an unkown service.", envType.Type)
+			return
 		}
 
+		env := &envelope{}
+		err = json.Unmarshal(b, env)
+		if err != nil {
+			log.Errorf("Unable to unmarshal message of type %v: %v", envType.Type, err)
+			return
+		}
+		// Pass this message and continue reading another one.
+		services[env.Type].in <- env.Message
```

Comment:
To elaborate on this further, since it's an important point ...

One way to think about this is that spawning goroutines to write to a channel is essentially creating an unbounded buffer.  Go will spawn as many goroutines as it can until it runs out of memory or cpu, and if nothing is reading from the channel (or it's reading very slowly), this virtual buffer will just increase in size until it exhausts the available resources.

As a general rule, in distributed systems I prefer to back-pressure once a reasonably sized buffer has been filled. If the UI is consistently producing messages faster than the backend can consume them, the backend should back pressure to the UI rather than trying to buffer until it exhausts its own resources.  That way, the UI will slow down (or start having errors), but the backend will remain stable and able to serve other consumers (e.g. user agents that are using the proxy).  The purpose of the buffer is just to absorb temporary shocks so that if the UI experiences temporarily high activity, we can absorb that without back pressuring.